File: app/src/main/java/com/wa2c/android/medoly/plugin/action/tweet/service/AbstractPluginService.java
package com.wa2c.android.medoly.plugin.action.tweet.service;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.text.TextUtils;

import com.twitter.twittertext.TwitterTextParseResults;
import com.twitter.twittertext.TwitterTextParser;
import com.wa2c.android.medoly.library.MediaPluginIntent;
import com.wa2c.android.medoly.library.PropertyData;
import com.wa2c.android.medoly.plugin.action.tweet.R;
import com.wa2c.android.medoly.plugin.action.tweet.util.Logger;
import com.wa2c.android.medoly.plugin.action.tweet.util.PropertyItem;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;



/**
 *  Plugin service base.
 */
public abstract class AbstractPluginService extends IntentService {

    /** Received receiver class name. */
    public static String RECEIVED_CLASS_NAME = "RECEIVED_CLASS_NAME";

    /**
     * Command result.
     */
    enum CommandResult {
        /** Succeeded. */
        SUCCEEDED,
        /** Failed. */
        FAILED,
        /** Authorization failed. */
        AUTH_FAILED,
        /** No media. */
        NO_MEDIA,
        /** Post saved. */
        SAVED,
        /** Ignore. */
        IGNORE
    }


    /** Context. */
    protected Context context = null;
    /** Preferences. */
    protected SharedPreferences sharedPreferences = null;
    /** Plugin intent. */
    protected MediaPluginIntent pluginIntent;
    /** Property data. */
    protected PropertyData propertyData;
    /** Received class name. */
    protected String receivedClassName;



    /**
     * Constructor.
     */
    public AbstractPluginService(String name) {
        super(name);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        Logger.d("onHandleIntent");
        if (intent == null)
            return;

        try {
            context = getApplicationContext();
            sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
            pluginIntent = new MediaPluginIntent(intent);
            propertyData = pluginIntent.getPropertyData();
            receivedClassName = pluginIntent.getStringExtra(RECEIVED_CLASS_NAME);
        } catch (Exception e) {
            Logger.e(e);
        }
    }

    @Override
    public void onDestroy(){
        super.onDestroy();
        Logger.d("onDestroy" + this.getClass().getSimpleName());
    }


    /**
     * Get tweet message text.
     * @return The message text.
     */
    protected String getTweetMessage() {
        // test data
        //propertyData.put(MediaProperty.TITLE, "０１２３４５６７８９０１２３４５６７８９０１２３４５６７８９０１２３４５６７８９０１２３４５６");
        //propertyData.put(MediaProperty.ARTIST, "０１２３４５６７８９０１２３４５６７８９０１２３４５６７８９０１２３４５６７８９");
        //propertyData.put(MediaProperty.ALBUM, "０１２３４５６７８９０１２３４５６７８９０１２３４５６７８９０１２３４５６７８９");

        final String format = sharedPreferences.getString(context.getString(R.string.prefkey_content_format), context.getString(R.string.format_content_default));
        final String TRIM_EXP = sharedPreferences.getBoolean(context.getString(R.string.prefkey_trim_before_empty_enabled), true) ? "\\w*" : "";
        List<PropertyItem> priorityList = PropertyItem.loadPropertyPriority(context);

        // get contains tag
        LinkedHashSet<PropertyItem> containsMap = new LinkedHashSet<>();
        for (PropertyItem item : priorityList) {
            Matcher matcher = Pattern.compile(item.getPropertyTag(), Pattern.MULTILINE).matcher(format);
            if (matcher.find())
                containsMap.add(item);
        }

        String outputText = format;
        for (PropertyItem propertyItem : containsMap) {
            String propertyText = propertyData.getFirst(propertyItem.propertyKey);
            String regexpText = propertyItem.getPropertyTag();
            if (TextUtils.isEmpty(propertyText)) {
                propertyText = "";
                regexpText = TRIM_EXP + regexpText;
            }

            String workText = outputText;
            Matcher matcher = Pattern.compile(regexpText).matcher(workText);
            while (matcher.find()) {
                workText = matcher.replaceFirst(propertyText);
                String removedText = getPropertyRemovedText(workText, containsMap);
                TwitterTextParseResults result = TwitterTextParser.parseTweet(removedText);
                int remainWeight = 999 - result.permillage; // 1000丁度だと入らない可能性
                if (remainWeight > 0) {
                    outputText = workText;
                } else {
                    if (propertyItem.shorten) {
                        workText = matcher.replaceFirst(trimWeightedText(propertyText, TwitterTextParser.parseTweet(propertyText).permillage + remainWeight));
                        outputText = getPropertyRemovedText(workText, containsMap);
                    }
                    break;
                }
            }
        }

        return outputText;
    }


    private String getPropertyRemovedText(String workText, Set<PropertyItem> containsMap) {
        String parseText = workText;
        for (PropertyItem pi : containsMap) {
            parseText = parseText.replaceAll("%" + pi.propertyKey + "%", "");
        }
        return parseText;
    }

    private String trimWeightedText(String propertyText, int remainWeight) {
        if (TextUtils.isEmpty(propertyText) || propertyText.length() < "...".length()) {
            return "";
        }

        Matcher newlineMatcher = Pattern.compile("\\r\\n|\\n|\\r").matcher(propertyText);
        if (newlineMatcher.find() && sharedPreferences.getBoolean(context.getString(R.string.prefkey_omit_newline), true)) {
            String returnText = "";
            while (newlineMatcher.find()) {
                TwitterTextParseResults result = TwitterTextParser.parseTweet(propertyText.substring(0, newlineMatcher.start()) + "...");
                if (result.permillage >= remainWeight) {
                    break;
                }
                returnText = propertyText.substring(0, newlineMatcher.start()) + "...";
            }
            return returnText;
        } else {
            for (int i = 1; i < propertyText.length(); i++) {
                TwitterTextParseResults result = TwitterTextParser.parseTweet(propertyText.substring(0, i) + "...");
                if (result.permillage >= remainWeight) {
                    return propertyText.substring(0, i - 1) + "...";
                }
            }
            return propertyText;
        }
    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/tweet/service/AbstractPluginService.kt
package com.wa2c.android.medoly.plugin.action.tweet.service

import android.annotation.SuppressLint
import android.app.IntentService
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.content.ContextCompat
import com.wa2c.android.medoly.library.MediaPluginIntent
import com.wa2c.android.medoly.library.PluginOperationCategory
import com.wa2c.android.medoly.library.PropertyData
import com.wa2c.android.medoly.plugin.action.tweet.R
import com.wa2c.android.medoly.plugin.action.tweet.util.AppUtils
import com.wa2c.android.prefs.Prefs
import timber.log.Timber
import java.io.InvalidObjectException


/**
 * Abstract plugin service.
 */
abstract class AbstractPluginService(name: String) : IntentService(name) {

    /** Context.  */
    protected lateinit var context: Context
    /** Preferences.  */
    protected lateinit var prefs: Prefs
    /** Plugin intent.  */
    protected lateinit var pluginIntent: MediaPluginIntent
    /** Property data.  */
    protected lateinit var propertyData: PropertyData
    /** Received class name.  */
    protected lateinit var receivedClassName: String
    /** Notification manager. */
    private var notificationManager : NotificationManager? = null


    @SuppressLint("NewApi")
    override fun onHandleIntent(intent: Intent?) {
        Timber.d("onHandleIntent")

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            val builder = Notification.Builder(this, NOTIFICATION_CHANNEL_ID)
                    .setContentTitle(getString(R.string.app_name))
                    .setSmallIcon(R.drawable.ic_notification)
            startForeground(NOTIFICATION_ID, builder.build())
        }

        if (intent == null)
            throw InvalidObjectException("Null intent")

        context = applicationContext
        prefs = Prefs(this)
        pluginIntent = MediaPluginIntent(intent)
        propertyData = pluginIntent.propertyData ?: PropertyData()
        receivedClassName = pluginIntent.getStringExtra(RECEIVED_CLASS_NAME)
    }

    override fun onDestroy() {
        super.onDestroy()
        notificationManager?.let {
            it.cancel(NOTIFICATION_ID)
            stopForeground(true)
        }
        Timber.d("onDestroy: %s", this.javaClass.simpleName)
    }


    /**
     * Show message.
     */
    fun showMessage(result: CommandResult, succeededMessage: String?, failedMessage: String?) {
        if (result == CommandResult.AUTH_FAILED) {
            AppUtils.showToast(context, R.string.message_account_not_auth)
        } else if (result == CommandResult.NO_MEDIA) {
            AppUtils.showToast(context, R.string.message_no_media)
        } else if (result == CommandResult.SUCCEEDED && !succeededMessage.isNullOrEmpty()) {
            if (pluginIntent.hasCategory(PluginOperationCategory.OPERATION_EXECUTE) || prefs.getBoolean(R.string.prefkey_tweet_success_message_show, defRes = R.bool.pref_default_tweet_success_message_show)) {
                AppUtils.showToast(context, succeededMessage)
            }
        } else if (result == CommandResult.FAILED && !failedMessage.isNullOrEmpty()) {
            if (pluginIntent.hasCategory(PluginOperationCategory.OPERATION_EXECUTE) || prefs.getBoolean(R.string.prefkey_tweet_failure_message_show, defRes = R.bool.pref_default_tweet_failure_message_show)) {
                AppUtils.showToast(context, failedMessage)
            }
        }
    }



    companion object {
        /** Notification ID */
        private const val NOTIFICATION_ID = 1
        /** Notification Channel ID */
        private const val NOTIFICATION_CHANNEL_ID = "Notification"

        /** Received receiver class name.  */
        const val RECEIVED_CLASS_NAME = "RECEIVED_CLASS_NAME"
        /** Previous data key.  */
        const val PREFKEY_PREVIOUS_MEDIA_URI = "previous_media_uri"

        /**
         * Create notification
         */
        fun createChannel(context: Context) {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O)
                return

            val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            if (notificationManager.getNotificationChannel(NOTIFICATION_CHANNEL_ID) == null) {
                val channel = NotificationChannel(NOTIFICATION_CHANNEL_ID, context.getString(R.string.app_name), NotificationManager.IMPORTANCE_MIN)
                notificationManager.createNotificationChannel(channel)
            }
        }
    }
}
